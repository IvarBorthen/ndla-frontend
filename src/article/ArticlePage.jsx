/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import React, { PropTypes, Component } from 'react';
import { connect } from 'react-redux';
import polyglot from '../i18n';

import * as actions from './articleActions';

class ArticlePage extends Component {
  componentWillMount() {
    const { fetchArticle } = this.props;
    fetchArticle('100');
  }

  render() {
    return (
      <div>
        <h1>{polyglot.t('hello.world')}</h1>
      </div>
    );
  }

}

ArticlePage.propTypes = {
  fetchArticle: PropTypes.func.isRequired,
};

const mapDispatchToProps = {
  fetchArticle: actions.fetchArticle,
};


export default connect(state => state, mapDispatchToProps)(ArticlePage);
