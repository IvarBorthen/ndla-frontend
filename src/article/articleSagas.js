/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { take, call, put, select } from 'redux-saga/effects';
import { getLocale } from '../locale/localeSelectors';
import { getArticle } from './articleSelectors';
import * as constants from './articleConstants';
import * as actions from './articleActions';
import * as api from './articleApi';

function* fetchArticle(id) {
  try {
    const locale = yield select(getLocale);
    const article = yield call(api.fetchArticle, id, locale);
    const articleWithId = Object.assign({}, article, { id });
    yield put(actions.setArticle(articleWithId));
  } catch (error) {
    // TODO: handle error
    // yield put(actions.applicationError());
  }
}

function* watchFetchArticle() {
  while (true) {
    const { payload: id } = yield take(constants.FETCH_ARTICLE);
    const currentArticle = yield select(getArticle);
    if (currentArticle.id !== id) {
      yield call(fetchArticle, id);
    }
  }
}

export default [
  watchFetchArticle,
];
