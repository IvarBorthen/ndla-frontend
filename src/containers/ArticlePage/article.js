/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { handleActions, createAction } from 'redux-actions';
import { createSelector } from 'reselect';
import { getLocale } from '../Locale/localeSelectors';
import { transformArticle } from '../../util/transformArticle';
import createFetchActions from '../../util/createFetchActions';

export const fetchArticleActions = createFetchActions('ARTICLE');
export const setArticle = createAction('SET_ARTICLE');
export const actions = {
  setArticle,
};

const initalState = {
  all: {},
  status: 'initial',
};

export default handleActions(
  {
    [actions.setArticle]: {
      next: (state, action) => ({
        ...state,
        all: { ...state.all, [action.payload.urn]: { ...action.payload } },
      }),
      throw: state => state,
    },
    [actions.fetchingArticle]: {
      next: state => ({
        ...state,
        status: 'loading',
      }),
      throw: state => state,
    },
    [actions.fetchArticleError]: {
      next: (state, action) => ({
        ...state,
        status:
          action.payload.error.json && action.payload.error.json.status === 404
            ? 'error404'
            : 'error',
      }),
      throw: state => state,
    },
    [actions.fetchArticleSuccess]: {
      next: state => ({
        ...state,
        status: 'success',
      }),
      throw: state => state,
    },
  },
  initalState,
);

const getArticlesFromState = state => state.articles;

export const getArticleByUrn = urn =>
  createSelector([getArticlesFromState], articles => articles.all[urn]);

export const getFetchStatus = state => state.articles.status;

export const getArticle = urn =>
  createSelector([getArticleByUrn(urn), getLocale], (article, locale) => {
    if (article) {
      return transformArticle(article, locale);
    }
    return undefined;
  });
