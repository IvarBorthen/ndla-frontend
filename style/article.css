.article {
  padding: 1em;
  max-width: 780px;
  margin: 0 auto;
}

.article img {
  max-width: 100%;
}


.figure_caption {
  display: block;
  font-style: italic;
  margin-bottom: 30px;
  font-family: 'Source Sans Pro', Helvetica, Arial, sans-serif;
  color: #4d4d4d;
  font-size: 0.75em;
  font-style: italic;
  font-weight: 400;
}

/** Content hacks (should be fixed in article-api or article-oembed)**/

/* Inngress? */
.article p strong {
  color: #4d4d4d;
  font-size: 1.2em;
  font-weight: 600;
}

.article strong em {
  font-style: normal;
}
