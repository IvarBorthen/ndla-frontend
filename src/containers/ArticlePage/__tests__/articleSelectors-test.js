/**
 * Copyright (c) 2016-present, NDLA.
 *
 * This source code is licensed under the GPLv3 license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

import { getArticle } from '../article';

const state = {
  locale: 'nb',
  articles: {
    all: {
      'urn:resource:1': {
        id: 1,
        urn: 'urn:resource:12',
        created: '2014-12-24T10:44:06Z',
        language: 'nb',
        title: 'Test resource 2',
        metaDescription: 'Beskrivelse',
        metaData: {},
      },
      'urn:topic:1': {
        id: 2,
        urn: 'urn:topic:123',
        created: '2014-11-24T10:44:06Z',
        language: 'en',
        title: 'Testing',
        metaDescription: 'Description',
        metaData: {},
      },
      'urn:resource:2': {
        id: 3,
        urn: 'urn:resource:2',
        created: '2014-12-24T10:44:06Z',
        updated: '2017-11-24T10:44:06Z',
        language: 'nb',
        title: 'Test resource 2',
        metaDescription: 'Beskrivelse',
        metaData: {},
      },
    },
  },
};

test('articleSelectors getArticle by urn', () => {
  expect(getArticle('urn:resource:1')(state)).toMatchSnapshot();
  expect(getArticle('urn:topic:123')(state)).toMatchSnapshot();
});

test('articleSelectors getArticle en locale', () => {
  const getArticleSelector = getArticle('urn:resource:2');
  const stateWithEnLocale = { ...state, locale: 'en' };
  expect(getArticleSelector(stateWithEnLocale)).toMatchSnapshot();
});

test('articleSelectors getArticle returns undefined if article is not in state', () => {
  const getArticleSelector = getArticle('urn:resource:1337');
  expect(getArticleSelector(state)).toEqual(undefined);
});
