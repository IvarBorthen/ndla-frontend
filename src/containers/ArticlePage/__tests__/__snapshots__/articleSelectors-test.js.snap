// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`articleSelectors getArticle by urn 1`] = `
Object {
  "created": "24.12.2014",
  "footNotes": Array [],
  "id": 1,
  "language": "nb",
  "metaData": Object {},
  "metaDescription": "Beskrivelse",
  "requiredLibraries": Array [],
  "title": "Test resource 2",
  "updated": "Invalid Date",
  "urn": "urn:resource:12",
}
`;

exports[`articleSelectors getArticle by urn 2`] = `undefined`;

exports[`articleSelectors getArticle en locale 1`] = `
Object {
  "created": "12/24/2014",
  "footNotes": Array [],
  "id": 3,
  "language": "nb",
  "metaData": Object {},
  "metaDescription": "Beskrivelse",
  "requiredLibraries": Array [],
  "title": "Test resource 2",
  "updated": "11/24/2017",
  "urn": "urn:resource:2",
}
`;
